File: src/main/services/XiaoyuzhouDetector.ts
import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';

/**
 * 小宇宙应用检测器
 * 自动检测 macOS 上小宇宙应用的安装位置和用户数据
 */
export interface XiaoyuzhouUserInfo {
  userId: string;
  dbPath: string;
  audioFilePath: string;
  userStoragePath: string;
}

export interface XiaoyuzhouAppInfo {
  containerId: string;
  containerPath: string;
  documentsPath: string;
  users: XiaoyuzhouUserInfo[];
}

export class XiaoyuzhouDetector {
  private readonly CONTAINER_BASE_PATH: string;
  private readonly KNOWN_CONTAINER_IDS = [
    // '8A51F41B-4985-4AD1-B6C2-384D6EC1A651', // 小宇宙已知的容器ID
  ];

  constructor() {
    const homeDir = os.homedir();
    this.CONTAINER_BASE_PATH = path.join(homeDir, 'Library', 'Containers');
  }

  /**
   * 自动检测小宇宙应用
   */
  async detectXiaoyuzhouApp(): Promise<XiaoyuzhouAppInfo | null> {
    try {
      // 1. 首先尝试已知的容器ID
      for (const containerId of this.KNOWN_CONTAINER_IDS) {
        const appInfo = await this.checkContainer(containerId);
        if (appInfo) {
          console.log(`✅ 找到小宇宙应用 (已知容器): ${containerId}`);
          return appInfo;
        }
      }

      // 2. 如果已知容器ID不存在，搜索所有容器
      console.log('🔍 在所有容器中搜索小宇宙应用...');
      const appInfo = await this.searchAllContainers();
      if (appInfo) {
        console.log(`✅ 找到小宇宙应用 (搜索发现): ${appInfo.containerId}`);
        return appInfo;
      }

      console.warn('⚠️  未找到小宇宙应用安装');
      return null;
    } catch (error: any) {
      console.error('检测小宇宙应用失败:', error);
      return null;
    }
  }

  /**
   * 检查指定容器ID
   */
  private async checkContainer(containerId: string): Promise<XiaoyuzhouAppInfo | null> {
    const containerPath = path.join(this.CONTAINER_BASE_PATH, containerId);
    
    if (!fs.existsSync(containerPath)) {
      return null;
    }

    const documentsPath = path.join(containerPath, 'Data', 'Documents');
    
    if (!fs.existsSync(documentsPath)) {
      return null;
    }

    // 检查是否存在 UserStorages 目录（小宇宙特征）
    const userStoragesPath = path.join(documentsPath, 'UserStorages');
    if (!fs.existsSync(userStoragesPath)) {
      return null;
    }

    // 获取所有用户
    const users = await this.getUsersFromDocuments(documentsPath);
    
    if (users.length === 0) {
      return null;
    }

    return {
      containerId,
      containerPath,
      documentsPath,
      users
    };
  }

  /**
   * 搜索所有容器
   */
  private async searchAllContainers(): Promise<XiaoyuzhouAppInfo | null> {
    try {
      if (!fs.existsSync(this.CONTAINER_BASE_PATH)) {
        return null;
      }

      const containers = fs.readdirSync(this.CONTAINER_BASE_PATH);
      
      for (const containerId of containers) {
        // 跳过已知的容器ID（已经检查过了）
        if (this.KNOWN_CONTAINER_IDS.includes(containerId)) {
          continue;
        }

        const appInfo = await this.checkContainer(containerId);
        if (appInfo) {
          return appInfo;
        }
      }

      return null;
    } catch (error) {
      console.error('搜索容器失败:', error);
      return null;
    }
  }

  /**
   * 从 Documents 目录获取用户列表
   */
  private async getUsersFromDocuments(documentsPath: string): Promise<XiaoyuzhouUserInfo[]> {
    const users: XiaoyuzhouUserInfo[] = [];

    try {
      // 方法1: 从 UserStorages 目录读取用户ID
      const userStoragesPath = path.join(documentsPath, 'UserStorages');
      if (fs.existsSync(userStoragesPath)) {
        const userDirs = fs.readdirSync(userStoragesPath).filter(dir => {
          const fullPath = path.join(userStoragesPath, dir);
          return fs.statSync(fullPath).isDirectory();
        });

        for (const userId of userDirs) {
          const userInfo = this.getUserInfo(documentsPath, userId);
          if (userInfo) {
            users.push(userInfo);
          }
        }
      }

      // 方法2: 从 Documents 根目录读取用户ID（直接存在用户目录）
      if (users.length === 0) {
        const entries = fs.readdirSync(documentsPath).filter(entry => {
          const fullPath = path.join(documentsPath, entry);
          if (!fs.statSync(fullPath).isDirectory()) {
            return false;
          }
          // 检查是否有 db/cosmos.db
          const dbPath = path.join(fullPath, 'db', 'cosmos.db');
          return fs.existsSync(dbPath);
        });

        for (const userId of entries) {
          const userInfo = this.getUserInfo(documentsPath, userId);
          if (userInfo) {
            users.push(userInfo);
          }
        }
      }

      return users;
    } catch (error) {
      console.error('获取用户列表失败:', error);
      return [];
    }
  }

  /**
   * 获取用户信息
   */
  private getUserInfo(documentsPath: string, userId: string): XiaoyuzhouUserInfo | null {
    // 用户存储路径可能在 Documents 根目录或 UserStorages 下
    const possiblePaths = [
      path.join(documentsPath, userId),
      path.join(documentsPath, 'UserStorages', userId)
    ];

    for (const userStoragePath of possiblePaths) {
      if (!fs.existsSync(userStoragePath)) {
        continue;
      }

      const dbPath = path.join(userStoragePath, 'db', 'cosmos.db');
      
      if (!fs.existsSync(dbPath)) {
        continue;
      }

      // 音频文件路径：/Documents/AudioFile/{userId}
      const audioFilePath = path.join(documentsPath, 'AudioFile', userId);

      return {
        userId,
        dbPath,
        audioFilePath,
        userStoragePath
      };
    }

    return null;
  }

  /**
   * 获取推荐的用户（如果只有一个用户，直接返回；否则返回第一个）
   */
  getRecommendedUser(appInfo: XiaoyuzhouAppInfo): XiaoyuzhouUserInfo | null {
    if (appInfo.users.length === 0) {
      return null;
    }

    // 如果只有一个用户，直接返回
    if (appInfo.users.length === 1) {
      console.log(`📌 自动选择唯一用户: ${appInfo.users[0].userId}`);
      return appInfo.users[0];
    }

    // 如果有多个用户，返回第一个（可以后续增加选择逻辑）
    console.log(`📌 发现 ${appInfo.users.length} 个用户，自动选择第一个: ${appInfo.users[0].userId}`);
    return appInfo.users[0];
  }

  /**
   * 格式化应用信息
   */
  formatAppInfo(appInfo: XiaoyuzhouAppInfo): string {
    const lines = [
      `容器ID: ${appInfo.containerId}`,
      `容器路径: ${appInfo.containerPath}`,
      `文档路径: ${appInfo.documentsPath}`,
      `用户数量: ${appInfo.users.length}`
    ];

    appInfo.users.forEach((user, index) => {
      lines.push(`\n用户 ${index + 1}:`);
      lines.push(`  - ID: ${user.userId}`);
      lines.push(`  - 数据库: ${user.dbPath}`);
      lines.push(`  - 音频文件: ${user.audioFilePath}`);
    });

    return lines.join('\n');
  }
}

File: src-tauri/src/services/xiaoyuzhou_detector.rs
use std::fs;
use std::path::{Path, PathBuf};

/// 小宇宙应用检测器
/// 自动检测 macOS 上小宇宙应用的安装位置和用户数据
pub struct XiaoyuzhouDetector;

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct XiaoyuzhouUserInfo {
    pub user_id: String,
    pub db_path: String,
    pub audio_file_path: String,
    pub user_storage_path: String,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct XiaoyuzhouAppInfo {
    pub container_id: String,
    pub container_path: String,
    pub documents_path: String,
    pub users: Vec<XiaoyuzhouUserInfo>,
}

impl XiaoyuzhouDetector {
    const KNOWN_CONTAINER_IDS: &'static [&'static str] = &[
        "8A51F41B-4985-4AD1-B6C2-384D6EC1A651",
    ];

    /// 获取容器基础路径
    fn container_base_path() -> PathBuf {
        let home = dirs::home_dir().unwrap_or_else(|| PathBuf::from("/"));
        home.join("Library").join("Containers")
    }

    /// 自动检测小宇宙应用
    pub fn detect() -> Option<XiaoyuzhouAppInfo> {
        // 1. 尝试已知的容器 ID
        for container_id in Self::KNOWN_CONTAINER_IDS {
            if let Some(app_info) = Self::check_container(container_id) {
                println!("✅ 找到小宇宙应用 (已知容器): {}", container_id);
                return Some(app_info);
            }
        }

        // 2. 搜索所有容器
        println!("🔍 在所有容器中搜索小宇宙应用...");
        Self::search_all_containers()
    }

    /// 检查指定容器 ID
    fn check_container(container_id: &str) -> Option<XiaoyuzhouAppInfo> {
        let container_path = Self::container_base_path().join(container_id);

        if !container_path.exists() {
            return None;
        }

        let documents_path = container_path.join("Data").join("Documents");
        if !documents_path.exists() {
            return None;
        }

        // 检查是否存在 UserStorages 目录
        let user_storages_path = documents_path.join("UserStorages");
        if !user_storages_path.exists() {
            return None;
        }

        let users = Self::get_users_from_documents(&documents_path);
        if users.is_empty() {
            return None;
        }

        Some(XiaoyuzhouAppInfo {
            container_id: container_id.to_string(),
            container_path: container_path.to_string_lossy().to_string(),
            documents_path: documents_path.to_string_lossy().to_string(),
            users,
        })
    }

    /// 搜索所有容器
    fn search_all_containers() -> Option<XiaoyuzhouAppInfo> {
        let base_path = Self::container_base_path();
        if !base_path.exists() {
            return None;
        }

        let entries = fs::read_dir(&base_path).ok()?;

        for entry in entries.flatten() {
            let container_id = entry.file_name().to_string_lossy().to_string();

            // 跳过已知的容器 ID
            if Self::KNOWN_CONTAINER_IDS.contains(&container_id.as_str()) {
                continue;
            }

            if let Some(app_info) = Self::check_container(&container_id) {
                println!("✅ 找到小宇宙应用 (搜索发现): {}", container_id);
                return Some(app_info);
            }
        }

        None
    }

    /// 从 Documents 目录获取用户列表
    fn get_users_from_documents(documents_path: &Path) -> Vec<XiaoyuzhouUserInfo> {
        let mut users = Vec::new();

        // 方法1: 从 UserStorages 目录读取
        let user_storages_path = documents_path.join("UserStorages");
        if user_storages_path.exists() {
            if let Ok(entries) = fs::read_dir(&user_storages_path) {
                for entry in entries.flatten() {
                    if entry.file_type().map(|t| t.is_dir()).unwrap_or(false) {
                        let user_id = entry.file_name().to_string_lossy().to_string();
                        if let Some(user_info) =
                            Self::get_user_info(documents_path, &user_id)
                        {
                            users.push(user_info);
                        }
                    }
                }
            }
        }

        // 方法2: 从 Documents 根目录读取
        if users.is_empty() {
            if let Ok(entries) = fs::read_dir(documents_path) {
                for entry in entries.flatten() {
                    if entry.file_type().map(|t| t.is_dir()).unwrap_or(false) {
                        let user_id = entry.file_name().to_string_lossy().to_string();
                        let db_path = entry.path().join("db").join("cosmos.db");
                        if db_path.exists() {
                            if let Some(user_info) =
                                Self::get_user_info(documents_path, &user_id)
                            {
                                users.push(user_info);
                            }
                        }
                    }
                }
            }
        }

        users
    }

    /// 获取用户信息
    fn get_user_info(documents_path: &Path, user_id: &str) -> Option<XiaoyuzhouUserInfo> {
        let possible_paths = [
            documents_path.join(user_id),
            documents_path.join("UserStorages").join(user_id),
        ];

        for user_storage_path in &possible_paths {
            if !user_storage_path.exists() {
                continue;
            }

            let db_path = user_storage_path.join("db").join("cosmos.db");
            if !db_path.exists() {
                continue;
            }

            let audio_file_path = documents_path.join("AudioFile").join(user_id);

            return Some(XiaoyuzhouUserInfo {
                user_id: user_id.to_string(),
                db_path: db_path.to_string_lossy().to_string(),
                audio_file_path: audio_file_path.to_string_lossy().to_string(),
                user_storage_path: user_storage_path.to_string_lossy().to_string(),
            });
        }

        None
    }

    /// 获取推荐用户
    pub fn get_recommended_user(app_info: &XiaoyuzhouAppInfo) -> Option<&XiaoyuzhouUserInfo> {
        app_info.users.first()
    }
}
